text/0000-inline-const.md

```diff
@@ -0,0 +1,276 @@
+>       | IfLetExpression
+>       | MatchExpression
+>    )
+>
+> ConstBlockExpression: `const` BlockExpression // new
+> ```
+
+This RFC extends the [grammar for patterns] to be,
+
+[grammar for patterns]: https://doc.rust-lang.org/stable/reference/patterns.html
+
+> ```
+> Pattern :
+>      LiteralPattern
+>    | IdentifierPattern
+>    | WildcardPattern
+>    | RangePattern
+>    | ReferencePattern
+>    | StructPattern
+>    | TupleStructPattern
+>    | TuplePattern
+>    | GroupedPattern
+>    | SlicePattern
+>    | PathPattern
+>    | MacroInvocation
+>    | ConstBlockExpression // new
+>
+> RangePatternBound :
+>      CHAR_LITERAL
+>    | BYTE_LITERAL
+>    | -? INTEGER_LITERAL
+>    | -? FLOAT_LITERAL
+>    | PathInExpression
+>    | QualifiedPathInExpression
+>    | ConstBlockExpression // new
+> ```
+
+In both the expression and pattern context, an inline `const` behaves exactly
+as if the user had declared a uniquely identified `const` with the block's
+contents as its initializer. For example, in expression context, writing
+`const { ... }` is equivalent to writing:
+
+```rust
+{ const UNIQUE_IDENT: Ty = ...; UNIQUE_IDENT }
+```
+
+where `Ty` is inferred from the expression inside the braces.
+
+An inline `const` is eligible for promotion in an implicit context (just like a
+named `const`), so the following are all guaranteed to work:
+
+```rust
+let x: &'static i32 = &const { 4i32.pow(4) }; // NOT IDIOMATIC
+let x: &'static i32 = const { &4i32.pow(4) }; // IDIOMATIC
+
+// If RFC 2203 is stabilized
+let v = [const { Vec::new() }; 3];  // IDIOMATIC
+let v = const { [ Vec::new(); 3] }; // NOT IDIOMATIC
+```
+
+Whether to lint against the non idiomatic versions is an open question.
+Personally, I would like to lint against `&const {...}` but not `const { [expr;
+3] }`.
+
+Inline `const`s are allowed within `const` and `static` initializers, just as we
+currently allow nested `const` declarations. Whether to lint against inline
+`const` expressions inside a `const` or `static` is also an open question.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This excludes other uses of the `const` keyword in expressions and patterns.
+I'm not aware of any other proposals that would take advantage of this.
+
+This would also be the first use of type inference for const initializers. I'm
+not aware of any technical issues that would arise from this, but perhaps I'm
+overlooking something?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative is the status quo. Maintaining it will likely result in
+promotion being used for more contexts. The lang-team decided to [explore this
+approach](https://github.com/rust-lang/rust/pull/70042#issuecomment-612221597)
+instead.
+
+It would also possible to separate out the parts of this RFC relating to patterns
+so that they can be decided upon seperately. I think they are similar enough
+that they are best considered as a unit, however.
```

Comment:
Given that so far we do not do implicit promotion in patterns, I feel it makes sense to go baby steps here and start with inline consts (anonymous consts?) as expressions only to match the existing implicit promotion in expressions -- so IMO they should be made separate feature gates and stabilization considered separately.

Reply:
For me, the overarching principles of this RFC should be:
1. inline consts behave exactly the same as if you defined a named constant in the same scope.
2. inline consts can be used anywhere you would use a named constant.

These rules of thumb help make this feature easy to reason about. Since you can already use named constants in patterns, I don't see any reason you shouldn't be allowed to use inline ones.

Reply:
The RFC describes inline consts as sugar for `{ const NAME := ...; NAME }`; that expansion would not work in patterns as they are not expressions.

I feel like patterns and expressions are sufficiently different that there is no reason to tie them together like that. For expressions we have concrete cases and this has been floated for years; for patterns this is the first time I see it and frankly I find it syntactically rather awkward.

Reply:
> The RFC describes inline consts as sugar for { const NAME := ...; NAME }; that expansion would not work in patterns as they are not expressions.

This is not news to me. The RFC states explicitly that that analogy is only valid in expression context. The real desugaring is less elegant–define a new constant with a unique name in the same scope as the anonymous `const` and then refer to the newly defined constant–but operationally the same. I don't think it's too big of a stretch to go from one to the other.

The rest is a matter of opinion. I find the consistency argument compelling, and I don't think wanting to match on the result of a `const fn` is all that novel, although admittedly my examples aren't very convincing.

Reply:
> The rest is a matter of opinion.

That is definitely so. :)
I am just asking that the RFC list arguments both way -- right now it lists your arguments but not mine.

Reply:
(From what I can see, the RFC still lists no arguments in favor of separating patterns.)